File: adapter/harness/preference_extractor.py
"""G-3 — FeedbackPreferenceExtractor.

Parses free-text user feedback into structured preference updates using
configurable keyword-to-field signal mappings.
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from typing import Any


@dataclass
class PreferenceSignal:
    patterns: list[str]
    field: str
    value: Any | None = None
    delta: float | None = None
    min_value: float | None = None
    max_value: float | None = None

    def __post_init__(self) -> None:
        if self.value is not None and self.delta is not None:
            raise ValueError("PreferenceSignal: value and delta are mutually exclusive; set at most one")


def make_preference_extractor(
    signals: list[PreferenceSignal],
    input_key: str = "feedback_text",
    output_key: str = "preference_updates",
    processed_flag_key: str = "feedback_processed",
) -> Callable[[dict], dict]:
    """Return a transform fn_ref-compatible function: (state: dict) -> dict."""

    def _extract(state: dict) -> dict:
        text = state.get(input_key)
        if not text:
            return state

        lower_text = text.lower()
        updates: dict[str, Any] = {}

        for signal in signals:
            if any(pattern.lower() in lower_text for pattern in signal.patterns):
                if signal.delta is not None:
                    base = state.get(signal.field, 0)
                    new_val = base + signal.delta
                    if signal.min_value is not None:
                        new_val = max(signal.min_value, new_val)
                    if signal.max_value is not None:
                        new_val = min(signal.max_value, new_val)
                    updates[signal.field] = new_val
                else:
                    updates[signal.field] = signal.value

        state[output_key] = updates
        state[processed_flag_key] = True
        return state

    return _extract

File: packages/harness/src/primitives/preference-extractor.ts
/**
 * G-3 — FeedbackPreferenceExtractor
 *
 * Parses free-text feedback into structured preference updates via
 * configurable keyword-to-field signal mappings.
 * Port of adapter/harness/preference_extractor.py. No domain vocabulary.
 *
 * Accepted as-is (plans/lexical_functions_hardening_plan.html Phase 3 step 5): this primitive is a
 * generic, caller-configured substring-matching engine with no content of its own. Any consumer
 * supplying `PreferenceSignal.patterns` inherits the same language/rephrasing limits as every other
 * lexical check in this plan — a future consumer shouldn't assume the primitive itself handles
 * non-English phrasing or paraphrase tolerance.
 */

export interface PreferenceSignal {
  patterns: string[]
  field: string
  value?: unknown
  delta?: number
  minValue?: number
  maxValue?: number
}

export function makePreferenceExtractor(
  signals: PreferenceSignal[],
  options?: {
    inputKey?: string
    outputKey?: string
    processedFlagKey?: string
  },
): (state: Record<string, unknown>) => Record<string, unknown> {
  for (const signal of signals) {
    if (signal.value != null && signal.delta != null) {
      throw new Error('PreferenceSignal: value and delta are mutually exclusive')
    }
  }

  const inputKey = options?.inputKey ?? 'feedback_text'
  const outputKey = options?.outputKey ?? 'preference_updates'
  const processedFlagKey = options?.processedFlagKey ?? 'feedback_processed'

  return function extract(state: Record<string, unknown>): Record<string, unknown> {
    const text = state[inputKey]
    if (!text) return state

    const lowerText = (text as string).toLowerCase()
    const updates: Record<string, unknown> = {}

    for (const signal of signals) {
      if (signal.patterns.some((p) => lowerText.includes(p.toLowerCase()))) {
        if (signal.delta != null) {
          const base = (state[signal.field] as number) ?? 0
          let newVal = base + signal.delta
          if (signal.minValue !== undefined) newVal = Math.max(signal.minValue, newVal)
          if (signal.maxValue !== undefined) newVal = Math.min(signal.maxValue, newVal)
          updates[signal.field] = newVal
        } else {
          updates[signal.field] = signal.value
        }
      }
    }

    return { ...state, [outputKey]: updates, [processedFlagKey]: true }
  }
}
